### DISCUSSION TITLE
Dynamic taints

### DISCUSSION THREAD

**Discussion by TodorSamardziski** [PyratLabs/ansible-role-k3s#143] [Q&A] [answered]
Hi,

Really complete role and easy to use ! 👍 

I can't add some taint to node after the playbook has been run one time. On first use :

```
k3s_server:
  node-taint:
    - "key1=value1:NoSchedule"
```

when run the playbook the first time node are created successfully with correct taint. And then the vars are changed to :

```
k3s_server:
  node-taint:
    - "key1=value1:NoSchedule"
    - "key2=value2:NoSchedule"
```

When we run the playbook again, the second taint isn't added.

Is it normal ? Or do we have to add some tasks to add / remove the taints dynamically ?

Best Regards.

**Comment by xanmanning** [comment, marked as answer]
No, this is not normal and not what I'd expect

This task writes the config file (which contains the taint): https://github.com/PyratLabs/ansible-role-k3s/blob/main/tasks/build/install-k3s-node.yml#L20-L28

When the file changes, [this notifies the handlers to restart k3s](https://github.com/PyratLabs/ansible-role-k3s/blob/main/tasks/build/install-k3s-node.yml#L25-L27).

Handlers are defined here: https://github.com/PyratLabs/ansible-role-k3s/blob/main/handlers/main.yml#L9-L21

Without further information, I can only assume that either:

1. It was run once successfully, it was run a second time unsuccessfully when the file changed so the handler wasn't run and then run a third/forth/etc time when the config file was not changed so the handler didn't execute.
2. `k3s_server` is being over-written in your config somewhere so the first configuration is in use.
3. You're caching facts.

For fixing these:

1. Restart k3s.
2. Only define `k3s_server` in one place
3. Clear Cache or don't cache facts when executing this role.

**Reply by TodorSamardziski** [reply]
Hi,

I'll check on my side, maybe I miss something. The configuration file  _/etc/rancher/k3s/config.yaml_ is changed successfully and k3s is restarted too. k3s_server var is only used in the group_vars. But the output of _kubectl describe node node1 | grep -i taint_ don't show that the new taint has been added. When trying to edit a node with "kubectl edit node node1" the second taint isn't present.  

I try to run the playbook with --flush-cache of course.

Thank you for your response, I'll check forward if something goes wrong.
